forte/train_preprocessor.py

```diff
@@ -0,0 +1,355 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from typing import Optional, Dict, Type, Any, Union, Iterator, List
+from texar.torch.data import DataIterator, Batch
+import torch
+from torch import device
+
+from forte.common.configuration import Config
+from forte.data.converter import Converter
+from forte.data.data_pack import DataPack
+from forte.data.data_pack_dataset import DataPackDataSource, \
+    DataPackDataset
+from forte.data.extractor.base_extractor import BaseExtractor
+from forte.data.ontology.core import Entry
+from forte.data.ontology.core import EntryType
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "TrainPreprocessor"
+]
+
+
+class TrainPreprocessor:
+    r"""
+    `TrainPreprocessor` provides the functionality of doing pre-processing work
+    including building vocabulary, extracting the features, batching and
+    padding (optional). The main functionality is provided by its method
+    :meth:`get_train_batch_iterator` which will return an `iterator` over the
+    batch of preprocessed data. Please refer to the documentation of
+    that method for how the pre-processing is done.
+
+    `TrainPreprocessor` will maintain a Config that stores all the configurable
+    parameters for various components.
+
+    `TrainPreprocessor` will also accept a user request. Internally it will
+    parse this user request and store the parsed result.
+
+    Args:
+        pack_generator (Iterator[DataPack]): A generator of
+            :class:`forte.data.data_pack.DataPack`.
+        request (Dict): A request that specifies how to do train pre-processing.
+            Please refer to :meth:`request` for details.
+        config: A `Dict` or :class:`forte.common.configuration.Config` that
+            configs this preprocessor. See :meth:`default_configs` for
+            the defaults.
+
+
+    .. note::
+        For parameters `request`, user does not necessarily need to provide
+        `converter`. If no `converter` is specified, a default converter of
+        type :class:`forte.data.converter.Converter` will be picked.
+    """
+
+    DATA_INPUT = 0
+    DATA_OUTPUT = 1
+
+    def __init__(self,
+                 pack_generator: Iterator[DataPack],
+                 request: Dict,
+                 config: Optional[Union[Config, Dict]] = None):
+        self._config: Config = \
+            Config(config, default_hparams=self.default_configs())
+        self._validate_config()
+
+        self._pack_generator: Iterator[DataPack] = pack_generator
+        self._cached_packs: List[DataPack] = []
+
+        self._user_request: Dict = request
+        self._request: Dict = {}
+        self._request_ready: bool = False
+        self._vocab_ready: bool = False
+
+        if not self._config.preprocess.lazy_parse_request:
+            self._parse_request(self._user_request)
```

Comment:
What's the benefit of parsing request lazily?

Reply:
I just would like to give users such flexibility. I guess there is no beneift currently. Probably I wil remove this option for now.

Reply:
Requests are simple json files right? I guess we don't need this option. It will somehow make people confused and will tend to think request parsing is an intensive job.

Reply:
OK